app/app.go

```diff
@@ -1394,6 +1364,45 @@ func (d *Deployments) GetDeploymentForDeviceWithCurrent(ctx context.Context, dev
 	return instructions, nil
 }
 
+func (d *Deployments) saveDeviceDeploymentRquest(ctx context.Context, deviceID string,
+	deviceDeployment *model.DeviceDeployment, request *model.DeploymentNextRequest) error {
+	if deviceDeployment.Request != nil {
+		if !reflect.DeepEqual(deviceDeployment.Request, request) {
+			// the device reported different device type and/or artifact name
+			// during the update process, which should never happen;
+			// mark deployment for this device as failed to force client to rollback
+			l := log.FromContext(ctx)
+			l.Errorf(
+				"Device with id %s reported new data: %s during update process;"+
+					"old data: %s",
+				deviceID, request, deviceDeployment.Request)
+
+			if err := d.UpdateDeviceDeploymentStatus(ctx, deviceDeployment.DeploymentId, deviceID,
+				model.DeviceDeploymentState{
+					Status: model.DeviceDeploymentStatusFailure,
```

Comment:
Random observation: I feel like we should set the SubState here to make it obvious from the UI what happened.

Reply:
The idea is valuable, but I don't think it fits into my PR.